src/merkle/tiered_smt/index.rs

```diff
@@ -0,0 +1,405 @@
+        Self { depth: 0, index: 0 }
+    }
+
+    // PROVIDERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Builds a node to be used as input of a hash function when computing a Merkle path.
+    ///
+    /// Will evaluate the parity of the current instance to define the result.
+    pub fn build_node(&self, slf: Word, sibling: Word) -> [RpoDigest; 2] {
+        if self.is_right_sibling() {
+            [sibling.into(), slf.into()]
+        } else {
+            [slf.into(), sibling.into()]
+        }
+    }
+
+    /// Returns the depth of the current instance.
+    pub const fn depth(&self) -> u8 {
+        self.depth
+    }
+
+    /// Returns the index of the current depth.
+    pub const fn index(&self) -> u64 {
+        self.index
+    }
+
+    /// Returns true if the current instance is at the maximum depth of `[TieredSmt]`.
+    pub const fn is_max_depth(&self) -> bool {
+        self.depth >= TieredSmt::MAX_DEPTH
+    }
+
+    /// Returns true if the current instance points to a right sibling node.
+    pub const fn is_right_sibling(&self) -> bool {
+        (self.index & 1) == 1
+    }
+
+    /// Returns `true` if the depth is `0`.
+    pub const fn is_root(&self) -> bool {
+        self.depth == 0
+    }
+
+    /// Computes the index of the sibling of the current node.
+    pub const fn sibling(mut self) -> Self {
+        self.index ^= 1;
+        self
+    }
+
+    /// Returns `true` if current depth is of a tier.
+    pub const fn is_tier(&self) -> bool {
+        (self.depth % TieredSmt::TIER_DEPTH) == 0
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Traverse towards the leaves, incrementing the depth by `1`.
+    pub fn traverse(&mut self, right: bool) -> &mut Self {
+        self.depth += 1;
+        self.index <<= 1;
+        self.index += right as u64;
+        self
+    }
+
+    /// Traverse one level towards the root, decrementing the depth by `1`.
+    pub fn backtrack(&mut self) -> &mut Self {
+        self.depth = self.depth.saturating_sub(1);
+        self.index >>= 1;
+        self
+    }
+
+    /// Traverse `n` levels towards the root.
+    pub fn backtrack_by(&mut self, n: u8) -> &mut Self {
+        self.depth = self.depth.saturating_sub(n);
+        self.index >>= n as u64;
+        self
+    }
+
+    /// Traverse to the previous tier.
+    pub fn backtrack_to_previous_tier(&mut self) -> &mut Self {
+        let n = (self.depth / TieredSmt::TIER_DEPTH) * TieredSmt::TIER_DEPTH;
+        let n = self.depth - n;
+        self.backtrack_by(n)
+    }
+}
+
+// LEAF INDEX
+// ================================================================================================
+
+/// A Merkle tree address to a tier node.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub struct LeafIndex {
+    key: CanonicalWord,
+    canonical: u64,
+    depth: u8,
+}
```

Comment:
This struct binds the state of a iteration, the `depth`, to a key. Removing it would increase code duplication and complexity as now we just call a couple of encapsulated methods to get all the things we need computed.

However, the `canonical` attribute could indeed be removed; fixed!